### PR TITLE
Make resources to block list configurable

### PR DESCRIPTION
**Pull request by xinyanw409**
**What this PR does / why we need it**:
Currently we hardcode the skip list as a map constant. If we want to edit skip crd resources list we need to build a new image every time. To make it more flexible, we're going to create a configmap for the skip list during backup-driver installation so we can add or delete skip list after plugin is deployed.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Tests done:
| Test case      | Result |
| ----------- | ----------- |
| Install velero vsphere plugin | ConfigMap velero-vsphere-plugin-block-list is created. |
| Edit configmap, then reboot velero pod | No new configmap created, the previous one keeps unchanged. |
| Delete configmap, then reboot velero pod | New configmap will be created from the default blocking list. |
| Delete configmap, then only reboot backup-driver pod | No new configmap created. |
| Backup without configmap | Succeed. |
| Restore using backup which created without configmap | Completed, but with 10 Warnings.|
| Backup with configmap, excluded resource `velero backup create test-0208-03 --include-namespaces=test-namespace-mapping-1 --exclude-resources=namespacenetworkinfos.nsx.vmware.com,contentsourcebindings.vmoperator.vmware.com,kubeadmconfigtemplates.bootstrap.cluster.x-k8s.io,clusterclasses.cluster.x-k8s.io,cnscsisvfeaturestates.cns.vmware.com --snapshot-volumes`| Succeed. |
| Restore using backup which created with configmap | Complete with 4 warnings. |
| Backup with configmap, not exclude resource `velero backup create test-0208-02 --include-namespaces=test-skip-list` | PartiallyFailed. 5 Errors. |

```
# kubectl get configmap velero-vsphere-plugin-block-list -n velero -o yaml
apiVersion: v1
data:
  agentinstalls.installers.tmc.cloud.vmware.com: "true"
  availabilityzones.topology.tanzu.vmware.com: "true"
  aviloadbalancerconfigs.netoperator.vmware.com: "true"
  backuprepositories.backupdriver.cnsdp.vmware.com: "true"
  backuprepositoryclaims.backupdriver.cnsdp.vmware.com: "true"
  blockaffinities.crd.projectcalico.org: "true"
  ccsplugins.appplatform.wcp.vmware.com: "true"
  certificaterequests.cert-manager.io: "true"
  certificates.cert-manager.io: "true"
  challenges.acme.cert-manager.io: "true"
  clonefromsnapshots.backupdriver.cnsdp.vmware.com: "true"
  clusterclasses.cluster.x-k8s.io: "true"
  clusterissuers.cert-manager.io: "true"
  clusternetworkinfos.nsx.vmware.com: "true"
  clusterresourcesetbindings.addons.cluster.x-k8s.io: "true"
  clusterresourcesets.addons.cluster.x-k8s.io: "true"
  clusters.cluster.x-k8s.io: "true"
  cnscsisvfeaturestates.cns.vmware.com: "true"
  cnsfileaccessconfigs.cns.vmware.com: "true"
  cnsfilevolumeclients.cns.vmware.com: "true"
  cnsnodevmattachments.cns.vmware.com: "true"
  cnsregistervolumes.cns.vmware.com: "true"
  cnsvolumemetadatas.cns.vmware.com: "true"
  compatibilities.run.tanzu.vmware.com: "true"
  contentlibraryproviders.vmoperator.vmware.com: "true"
  contentsourcebindings.vmoperator.vmware.com: "true"
  contentsources.vmoperator.vmware.com: "true"
  deletesnapshots.backupdriver.cnsdp.vmware.com: "true"
  downloads.datamover.cnsdp.vmware.com: "true"
  drainrequests.psp.wcp.vmware.com: "true"
  gatewayclasses.networking.x-k8s.io: "true"
  gateways.networking.x-k8s.io: "true"
  haproxyloadbalancerconfigs.netoperator.vmware.com: "true"
  httproutes.networking.x-k8s.io: "true"
  imagedisks.imagecontroller.vmware.com: "true"
  installoptions.appplatform.wcp.vmware.com: "true"
  installrequirements.appplatform.wcp.vmware.com: "true"
  ipamblocks.crd.projectcalico.org: "true"
  ipamconfigs.crd.projectcalico.org: "true"
  ipamhandles.crd.projectcalico.org: "true"
  ippools.crd.projectcalico.org: "true"
  ippools.netoperator.vmware.com: "true"
  ippools.nsx.vmware.com: "true"
  issuers.cert-manager.io: "true"
  kubeadmconfigs.bootstrap.cluster.x-k8s.io: "true"
  kubeadmconfigtemplates.bootstrap.cluster.x-k8s.io: "true"
  kubeadmcontrolplanes.controlplane.cluster.x-k8s.io: "true"
  kuberneteslicenses.licenseoperator.vmware.com: "true"
  loadbalancerconfigs.netoperator.vmware.com: "true"
  loadbalancers.vmware.com: "true"
  machinedeployments.cluster.x-k8s.io: "true"
  machinehealthchecks.cluster.x-k8s.io: "true"
  machinepools.exp.cluster.x-k8s.io: "true"
  machines.cluster.x-k8s.io: "true"
  machinesets.cluster.x-k8s.io: "true"
  members.registryagent.vmware.com: "true"
  namespacenetworkinfos.nsx.vmware.com: "true"
  ncpconfigs.nsx.vmware.com: "true"
  networkinterfaces.netoperator.vmware.com: "true"
  networks.netoperator.vmware.com: "true"
  nsxerrors.nsx.vmware.com: "true"
  nsxlocks.nsx.vmware.com: "true"
  nsxnetworkconfigurations.nsx.vmware.com: "true"
  nsxnetworkinterfaces.nsx.vmware.com: "true"
  orders.acme.cert-manager.io: "true"
  persistenceinstanceinfoes.psp.wcp.vmware.com: "true"
  persistenceserviceconfigurations.psp.wcp.vmware.com: "true"
  projects.registryagent.vmware.com: "true"
  providerserviceaccounts.run.tanzu.vmware.com: "true"
  psprecencyrequests.psp.wcp.vmware.com: "true"
  pvcdisruptionbudgets.psp.wcp.vmware.com: "true"
  registries.registryagent.vmware.com: "true"
  resourcecheckreports.psp.wcp.vmware.com: "true"
  resourcechecks.psp.wcp.vmware.com: "true"
  routesets.nsx.vmware.com: "true"
  snapshots.backupdriver.cnsdp.vmware.com: "true"
  statefuldrainnodes.psp.wcp.vmware.com: "true"
  statefulreadynodes.psp.wcp.vmware.com: "true"
  storagepolicies.appplatform.wcp.vmware.com: "true"
  storagepolicies.psp.wcp.vmware.com: "true"
  storagepools.cns.vmware.com: "true"
  supervisorservices.appplatform.wcp.vmware.com: "true"
  tanzukubernetesaddons.run.tanzu.vmware.com: "true"
  tanzukubernetesclusters.run.tanzu.vmware.com: "true"
  tanzukubernetesreleases.run.tanzu.vmware.com: "true"
  tcproutes.networking.x-k8s.io: "true"
  tkgserviceconfigurations.run.tanzu.vmware.com: "true"
  uploads.datamover.cnsdp.vmware.com: "true"
  vcuiplugins.appplatform.wcp.vmware.com: "true"
  veleroservices.veleroappoperator.vmware.com: "true"
  virtualmachineclassbindings.vmoperator.vmware.com: "true"
  virtualmachineclasses.vmoperator.vmware.com: "true"
  virtualmachineimages.vmoperator.vmware.com: "true"
  virtualmachines.vmoperator.vmware.com: "true"
  virtualmachineservices.vmoperator.vmware.com: "true"
  virtualmachinesetresourcepolicies.vmoperator.vmware.com: "true"
  virtualnetworkinterfaces.vmware.com: "true"
  virtualnetworks.vmware.com: "true"
  vmxnet3networkinterfaces.netoperator.vmware.com: "true"
  vspheredistributednetworks.netoperator.vmware.com: "true"
  wcpclusters.infrastructure.cluster.vmware.com: "true"
  wcpmachines.infrastructure.cluster.vmware.com: "true"
  wcpmachinetemplates.infrastructure.cluster.vmware.com: "true"
  wcpnamespaces.appplatform.wcp.vmware.com: "true"
  webconsolerequests.vmoperator.vmware.com: "true"
kind: ConfigMap
metadata:
  creationTimestamp: "2023-01-26T02:59:41Z"
  name: velero-vsphere-plugin-block-list
  namespace: velero
  resourceVersion: "4828873"
  uid: 3f0d70e3-787a-46f7-b552-428da16729c0
```
2. Unit test
```
go test ./pkg/... -timeout=300s
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/apis/backupdriver/v1alpha1	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/apis/datamover/v1alpha1	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/backupdriver	[no test files]
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/backuprepository	0.217s
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/builder	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/buildinfo	[no test files]
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/cmd	0.224s
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/cmd/backupdriver	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/cmd/backupdriver/cli/install	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/cmd/backupdriver/cli/server	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/cmd/datamgr	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/cmd/datamgr/cli/install	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/cmd/datamgr/cli/server	[no test files]
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/common/vsphere	0.115s
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/constants	[no test files]
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/controller	0.231s
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/dataMover	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/clientset/versioned	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/clientset/versioned/fake	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/clientset/versioned/scheme	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/clientset/versioned/typed/backupdriver/v1alpha1	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/clientset/versioned/typed/backupdriver/v1alpha1/fake	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/clientset/versioned/typed/datamover/v1alpha1	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/clientset/versioned/typed/datamover/v1alpha1/fake	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/crds	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/informers/externalversions	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/informers/externalversions/backupdriver	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/informers/externalversions/backupdriver/v1alpha1	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/informers/externalversions/datamover	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/informers/externalversions/datamover/v1alpha1	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/informers/externalversions/internalinterfaces	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/listers/backupdriver/v1alpha1	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/generated/listers/datamover/v1alpha1	[no test files]
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/install	[no test files]
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/ivd	0.086s
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/paravirt	0.130s
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin	[no test files]
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/plugin/util	0.117s
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotUtils	0.139s
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr	0.069s
?   	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/test	[no test files]
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/utils	0.062s
Success!
```
**Does this PR introduce a user-facing change?**:

```release-note
Make resources to block list configurable.
```

### CODE REVIEW

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/778/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/779/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/780/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/781/

File: Dockerfile-datamgr [hunk 1]
```diff
@@ -15,7 +15,7 @@
 FROM busybox:1.33.1 AS busybox
 
 FROM photon:4.0
-ADD /bin/linux/amd64/data-* /datamgr
+ADD /bin/linux/amd64/data-* /datamgr/
```
Comment (xing-yang) [inline Dockerfile-datamgr:18]: Is this change needed?

File: pkg/cmd/utils.go [hunk 2]
```diff
@@ -254,3 +254,77 @@ func CheckPluginImageRepo(kubeClient kubernetes.Interface, ns string, defaultIma
...
+				Namespace: veleroNs,
+			},
+			Data: make(map[string]string),
+		}
+		create = true
+	}
+
+	// If there is blocking list configmap created already, leave the previous one unchanged.
+	if !create {
+		return nil
```
Comment (deepakkinni) [inline pkg/cmd/utils.go:282]: Can you add a log statement that states that we are re-using older configmap.

File: pkg/cmd/utils.go [hunk 3]
```diff
@@ -254,3 +254,77 @@ func CheckPluginImageRepo(kubeClient kubernetes.Interface, ns string, defaultIma
...
+	}
+
+	blockListData := make(map[string]string)
+	for resourceToBlock, isBlock := range resourceToBlock {
+		blockListData[resourceToBlock] = strconv.FormatBool(isBlock)
+	}
+	blockListConfigMap.Data = blockListData
+	_, err = kubeClient.CoreV1().ConfigMaps(veleroNs).Create(ctx, blockListConfigMap, metav1.CreateOptions{})
+	if err != nil {
+		fmt.Printf("Failed to create/update resources to block lost config map : %s.\n", constants.ResourcesToBlockList)
```
Comment (deepakkinni) [inline pkg/cmd/utils.go:292]: "Failed"->"failed"

File: pkg/constants/constants.go [hunk 4]
```diff
@@ -187,6 +187,11 @@ const (
 	SupervisorResourcePoolKey = "SupervisorResourcePool"
 )
 
+// Name for skip resources list configmap
+const (
+	ResourcesToBlockList = "velero-vsphere-plugin-block-list"
```
Comment (deepakkinni) [inline pkg/constants/constants.go:192]: Can you change the variable to indicate that it is a name and not a list.

File: pkg/cmd/utils.go [hunk 5]
```diff
@@ -254,3 +254,77 @@ func CheckPluginImageRepo(kubeClient kubernetes.Interface, ns string, defaultIma
 
 	return resultImage, err
 }
+
+
+// If there is no configmap velero-vsphere-plugin-block-list created before, create a new configmap from the default blocking list.
+// If there is already a configmap velero-vsphere-plugin-block-list created, leave the previous one unchanged.
+func CreateBlockListConfigMap(kubeClient kubernetes.Interface, veleroNs string, resourceToBlock map[string]bool) error {
```
Comment (xing-yang) [inline pkg/cmd/utils.go:261]: What if user deleted the ConfigMap and rebooted the plugin?

File: pkg/cmd/utils.go [hunk 6]
```diff
@@ -254,3 +254,77 @@ func CheckPluginImageRepo(kubeClient kubernetes.Interface, ns string, defaultIma
...
+	}
+
+	blockListConfigMap, err := kubeClient.CoreV1().ConfigMaps(veleroNs).Get(ctx, constants.ResourcesToBlockList, metav1.GetOptions{})
+	if err != nil {
+		var errMsg string
+		if !k8serrors.IsNotFound(err) {
+			errMsg = fmt.Sprintf("Failed to retrieve configmap: %s.\n", constants.ResourcesToBlockList)
+		} else {
+			errMsg = fmt.Sprintf("Failed to find configmap: %s. \n", constants.ResourcesToBlockList)
+		}
```
Comment (deepakkinni) [inline pkg/cmd/utils.go:325]: Net effect seems the same irrespective of the type of error, so instead just have a generic err message but log the err.

File: Dockerfile-datamgr (hunk 1 above)
Comment (xinyanw409) [reply Dockerfile-datamgr:18]: Without this change build will failed with error "When using ADD with more than one source file, the destination must be a directory and end with a /".

[the remark this answers is not in the document]
Comment (xinyanw409) [reply Dockerfile-datamgr:18]: Same as above.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/782/

File: pkg/cmd/utils.go (hunk 5 above)
Comment (xinyanw409) [reply pkg/cmd/utils.go:261]: This logic is called during starting init container. If user deletes the ConfigMap, then reboot the plugin(scale down and up pods), the ConfigMap will be recreated during starting the init containers.

This is only executed once during initialing containers, for every single backup, it will retrieve the existed configmap, if no configmap found, assuming there is no resource to be blocked.

File: pkg/plugin/util/util.go [hunk 7]
```diff
@@ -284,25 +286,31 @@ func IsObjectBlocked(item runtime.Unstructured) (bool, string, error) {
...
+	for resourceToBlock, _ := range blockListConfigMap {
 		desiredResources = append(desiredResources, resourceToBlock)
 	}
 	return desiredResources
 }
 
-func IsResourceBlocked(resourceName string) bool {
-	return constants.ResourcesToBlock[resourceName]
+func IsResourceBlocked(resourceName string, resourcesToBlock map[string]string) bool {
+	isBlocked, _ := strconv.ParseBool(resourcesToBlock[resourceName])
```
Comment (deepakkinni) [inline pkg/plugin/util/util.go:312]: `resourcesToBlock` could be nil

File: pkg/plugin/util/util.go [hunk 8]
```diff
@@ -284,25 +286,31 @@ func IsObjectBlocked(item runtime.Unstructured) (bool, string, error) {
...
+	if IsResourceBlocked(crdName, blockListConfigMap) {
 		return true, crdName, nil
 	}
 	return false, crdName, nil
 }
 
-func GetResources() []string {
-	desiredResources := make([]string, len(constants.ResourcesToHandle)+len(constants.ResourcesToBlock))
+func GetResources(blockListConfigMap map[string]string) []string {
+	desiredResources := make([]string, len(constants.ResourcesToHandle)+len(blockListConfigMap))
```
Comment (deepakkinni) [inline pkg/plugin/util/util.go:301]: `blockListConfigMap` could be nil, add unit test for this scenario

**Comment by deepakkinni** [comment]
Update description to indicate exactly when the configmap will be recreated and when it won't be

File: pkg/plugin/util/util.go (hunk 7 above)
Comment (xinyanw409) [reply pkg/plugin/util/util.go:312]: Added unit test for nil case. Golang can handle this.

File: pkg/plugin/util/util.go (hunk 8 above)
Comment (xinyanw409) [reply pkg/plugin/util/util.go:301]: Line 40: check if error type is k8serrors.IsNotFound. 
If !k8serrors.IsNotFound, return error.
if k8serrors.IsNotFound, 
line 45: blockListConfigMap = make(map[string]string)
so in this case blockListConfigMap will be empty but not nil.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/783/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/784/

**Review by deepakkinni — APPROVED** [review]

**Comment by xing-yang** [comment]
I wonder why backup results from (1) "Backup without configmap" and (2) "Backup with configmap, not exclude resource" are so different.  They should be the same.  Are you running backup test (2) from a restored setup resulting from (1)?

**Comment by xinyanw409** [comment]
> I wonder why backup results from (1) "Backup without configmap" and (2) "Backup with configmap, not exclude resource" are so different. They should be the same. Are you running backup test (2) from a restored setup resulting from (1)?

- They should be the same.
But "Backup with configmap, not exclude resource" should failed because not passing "--exclude-resource" and there are resources are in  block list. 

- Are you running backup test (2) from a restored setup resulting from (1)?
No.  I created two namespaces and created the pvc in the namespace.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/785/

**Review by xing-yang — APPROVED** [review]
LGTM
